src/testing/vortex/supervisor.zig

```diff
@@ -0,0 +1,336 @@
+        // required duration.
+        log.info("terminating workload due to max duration", .{});
+        break :blk if (workload.state() == .running)
+            try workload.terminate(std.posix.SIG.KILL)
+        else
+            try workload.wait();
+    };
+
+    switch (workload_result) {
+        .Signal => |signal| {
+            switch (signal) {
+                std.posix.SIG.KILL => log.info(
+                    "workload terminated as requested",
+                    .{},
+                ),
+                else => {
+                    log.err(
+                        "workload exited unexpectedly with signal {d}",
+                        .{signal},
+                    );
+                    process.exit(1);
+                },
+            }
+        },
+        else => {
+            log.err("unexpected workload result: {any}", .{workload_result});
+            return error.TestFailed;
+        },
+    }
+}
+
+fn terminate_random_replica(random: std.Random, replicas: []*Replica) !void {
+    if (random_replica_in_state(random, replicas, .running)) |found| {
+        log.info("terminating replica {d}", .{found.index});
+        _ = try found.replica.terminate();
+        log.info("replica {d} terminated", .{found.index});
+    } else return error.NoRunningReplica;
+}
+
+const RandomReplica = struct { replica: *Replica, index: u8 };
+
+fn random_replica_in_state(
+    random: std.Random,
+    replicas: []*Replica,
+    state: Replica.State,
+) ?RandomReplica {
+    var matching: [replica_count]RandomReplica = undefined;
+    var count: u8 = 0;
+
+    for (replicas, 0..) |replica, index| {
+        if (replica.state() == state) {
+            matching[count] = .{ .replica = replica, .index = @intCast(index) };
+            count += 1;
+        }
+    }
+    if (count == 0) {
+        return null;
+    } else {
+        return matching[random.uintLessThan(usize, count)];
+    }
+}
+
+fn start_workload(allocator: std.mem.Allocator) !*LoggedProcess {
+    const argv = &.{
+        "zig-out/bin/vortex",
+        "workload",
+        std.fmt.comptimePrint("--cluster-id={d}", .{cluster_id}),
+        std.fmt.comptimePrint("--addresses={s}", .{replica_addresses}),
+        "--driver-command=zig-out/bin/vortex driver",
+    };
+
+    return try LoggedProcess.spawn(allocator, argv);
+}
+
+pub fn comma_separate_ports(comptime ports: []const u16) []const u8 {
+    assert(ports.len > 0);
+
+    var out: []const u8 = std.fmt.comptimePrint("{d}", .{ports[0]});
+
+    inline for (ports[1..]) |port| {
+        out = out ++ std.fmt.comptimePrint(",{d}", .{port});
+    }
+
+    assert(std.mem.count(u8, out, ",") == ports.len - 1);
+    return out;
+}
+
+// Create a new Vortex-specific temporary directory. Caller owns the returned memory.
+fn create_tmp_dir(allocator: std.mem.Allocator) ![]const u8 {
+    const root = process.getEnvVarOwned(allocator, "TMPDIR") catch try allocator.dupe(u8, "/tmp");
```

Comment:
Let's use `.zig-cache/tmp` for our temp dir. I think we can assume that supervisor is being run with cwd pointing at the repository root? 

That way, we don't depend on ambient environment (TMPDIR) and actually keep stuff scoped to the project. And other things already use zig-cache/tmp, so we also gain uniformity here.

Reply:
I think if someone uses vortex outside of the repo, they could live with `.zig-cache` created in the cwd?